# __Dropout Regularization__

### This notebook demonstrates dropout regularization implementation in neural networks using TensorFlow/Keras.

  Key components:
  - Uses a customer dataset with demographic/behavioral features to predict product preferences (7 categories)
  - Implements a 3-layer neural network with 50% dropout on the first hidden layer and L2 regularization on the second layer
  - Performs 5-fold cross-validation showing fold accuracies

Dropout is a technique where:

- Randomly selected neurons are ignored during training. They are 'dropped out' randomly. This means that their contribution to the activation of downstream neurons is temporally removed on the forward pass, and any weight updates are not applied to the neuron on the backward pass.


- If neurons are randomly dropped out of the network during training, other neurons will have to step in and handle the representation required to make predictions for the missing neurons. This is believed to result in multiple independent internal representations being learned by the network.

- The effect is that the network becomes less sensitive to the specific weights of neurons. This, in turn, results in a network that is capable of better generalization and is less likely to overfit the training data.

Let's understand how it works.


## Steps to be followed:
1. Import the required libraries
2. Read a CSV file into a DataFrame
3. Create dummies
4. Standardize and prepare data for modeling
5. Perform K-fold cross-validation and model training
6. Final accuracy calculation on the test set

### Step 1: Import the required libraries

- Import libraries for data preprocessing, including z-score standardization using **scipy.stats.zscore** and data manipulation using **pandas**. It also imports libraries for model evaluation, such as metrics from **sklearn** and train-test splitting from **sklearn.model_selection**.
- Import the necessary components from TensorFlow Keras (**Sequential** and **Dense**) to build a neural network model. These components allow for the creation of a sequential model with dense layers and activation functions.

In [1]:
# !pip install tensorflow==2.17.0 scikeras==0.13.0 keras==3.2.0

In [2]:
import os

# Disable oneDNN optimizations to avoid potential minor numerical differences caused by floating-point round-off errors.
os.environ['TF_ENABLE_ONEDNN_OPTS'] = '0'

In [6]:
import pandas as pd
import numpy as np
from scipy.stats import zscore
import numpy as np
import tensorflow as tf
import sklearn
from sklearn import metrics
from sklearn.model_selection import train_test_split, KFold
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras import regularizers
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
import scikeras

print("TensorFlow version:", tf.__version__)
print("Keras version:", tf.keras.__version__)
print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)
print("SciPy version:", pd.__version__)
print("Scikit-learn version:", sklearn.__version__)
print("Scikeras version:", scikeras.__version__)


TensorFlow version: 2.17.1
Keras version: 3.11.3
Pandas version: 2.2.3
NumPy version: 1.26.4
SciPy version: 2.2.3
Scikit-learn version: 1.4.2
Scikeras version: 0.13.0


### Step 2: Read a CSV file into a DataFrame
- Read a CSV file from a given URL and stores it in a Pandas DataFrame by using **na_values** to replace specified values with missing values.

In [7]:
# Read the data set
df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/jh-simple-dataset.csv",
    na_values=['NA','?'])

## Data Description

### **Column Details:**
- `id`: Unique identifier for each row
- `job`: Categorical variable (33 unique job codes)
- `area`: Categorical variable (4 distinct areas)
- `income`: Numeric, annual income (some missing values)
- `aspect`: Numeric, possibly a survey or rating value
- `subscriptions`: Integer count of subscriptions (0–5 typical range)
- `dist_healthy`: Distance to healthy food options (continuous)
- `save_rate`: Savings rate (contains unusual outliers: negative and very high values)
- `dist_unhealthy`: Distance to unhealthy food options (continuous)
- `age`: Age of individual (29–52 years) → **target variable**
- `pop_dense`: Normalized population density (0–1 scale)
- `retail_dense`: Normalized retail density (0–1 scale)
- `crime`: Normalized crime index (0–1 scale)
- `product`: Categorical variable with 7 unique product categories

### **Notes**
- This dataset mixes **categorical** (job, area, product) and **numerical** variables.
- `income` has missing values that need handling.
- `save_rate` includes outliers (−119 to 384) → potential data quality issue.
- `age` is the dependent variable, making this a **regression task**.


In [8]:
df.head(5)

id job area   income     aspect  subscriptions  dist_healthy  save_rate  \
0   1  vv    c  50876.0  13.100000              1      9.017895         35   
1   2  kd    c  60369.0  18.625000              2      7.766643         59   
2   3  pe    c  55126.0  34.766667              1      3.632069          6   
3   4  11    c  51690.0  15.808333              1      5.372942         16   
4   5  kl    d  28347.0  40.941667              3      3.822477         20   

   dist_unhealthy  age  pop_dense  retail_dense     crime product  
0       11.738935   49   0.885827      0.492126  0.071100       b  
1        6.805396   51   0.874016      0.342520  0.400809       c  
2       13.671772   44   0.944882      0.724409  0.207723       b  
3        4.333286   50   0.889764      0.444882  0.361216       b  
4        5.967121   38   0.744094      0.661417  0.068033       a

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              2000 non-null   int64  
 1   job             2000 non-null   object 
 2   area            2000 non-null   object 
 3   income          1941 non-null   float64
 4   aspect          2000 non-null   float64
 5   subscriptions   2000 non-null   int64  
 6   dist_healthy    2000 non-null   float64
 7   save_rate       2000 non-null   int64  
 8   dist_unhealthy  2000 non-null   float64
 9   age             2000 non-null   int64  
 10  pop_dense       2000 non-null   float64
 11  retail_dense    2000 non-null   float64
 12  crime           2000 non-null   float64
 13  product         2000 non-null   object 
dtypes: float64(7), int64(4), object(3)
memory usage: 218.9+ KB


In [11]:
df.describe()

id        income       aspect  subscriptions  dist_healthy  \
count  2000.000000   1941.000000  2000.000000    2000.000000   2000.000000   
mean   1000.500000  56924.861927    21.135679       1.199000      9.295306   
std     577.494589  10188.938399    12.088238       0.954911      5.731798   
min       1.000000  19846.000000     4.000000       0.000000      2.000000   
25%     500.750000  51336.000000    10.608333       1.000000      4.937724   
50%    1000.500000  58633.000000    18.083333       1.000000      8.065855   
75%    1500.250000  64407.000000    32.925000       2.000000     12.227631   
max    2000.000000  88671.000000    42.891667       5.000000     33.009306   

         save_rate  dist_unhealthy          age    pop_dense  retail_dense  \
count  2000.000000     2000.000000  2000.000000  2000.000000   2000.000000   
mean     47.549500       14.316942    45.836500     0.879758      0.563258   
std      58.177598        8.209951     3.703867     0.076570      0.152759   
min    -119.000000        2.160180    29.000000     0.492126      0.000000   
25%       7.000000        7.858578    44.000000     0.838583      0.472441   
50%      31.000000       12.376985    47.000000     0.889764      0.562992   
75%      70.000000       19.312772    49.000000     0.933071      0.673228   
max     384.000000       37.052691    52.000000     1.000000      0.972441   

             crime  
count  2000.000000  
mean      0.275939  
std       0.182886  
min       0.000000  
25%       0.141712  
50%       0.241322  
75%       0.356615  
max       0.977276

In [12]:
# Check for missing values
df.isnull().sum()

id                 0
job                0
area               0
income            59
aspect             0
subscriptions      0
dist_healthy       0
save_rate          0
dist_unhealthy     0
age                0
pop_dense          0
retail_dense       0
crime              0
product            0
dtype: int64

#### Create dummy variables

- Use the **pd.get_dummies()** function to convert categorical columns **'job'** and **'area'** into dummy variables, which represent the presence or absence of each category as binary values.

- Drop the original categorical columns **'job'** and **'area'** from the DataFrame using the **df.drop()** function

In [13]:
# Convert categorical columns to dummy variables
categorical_cols = ['job', 'area']
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

In [14]:
df.columns


Index(['id', 'income', 'aspect', 'subscriptions', 'dist_healthy', 'save_rate',
       'dist_unhealthy', 'age', 'pop_dense', 'retail_dense', 'crime',
       'product', 'job_al', 'job_am', 'job_ax', 'job_bf', 'job_by', 'job_cv',
       'job_de', 'job_dz', 'job_e2', 'job_f8', 'job_gj', 'job_gv', 'job_kd',
       'job_ke', 'job_kl', 'job_kp', 'job_ks', 'job_kw', 'job_mm', 'job_nb',
       'job_nn', 'job_ob', 'job_pe', 'job_po', 'job_pq', 'job_pz', 'job_qp',
       'job_qw', 'job_rn', 'job_sa', 'job_vv', 'job_zz', 'area_b', 'area_c',
       'area_d'],
      dtype='object')

In [15]:
# Convert all boolean columns to numeric (integer) format
boolean_columns = df.select_dtypes(include=['bool']).columns
df[boolean_columns] = df[boolean_columns].astype(int)

In [16]:
df.head()

id   income     aspect  subscriptions  dist_healthy  save_rate  \
0   1  50876.0  13.100000              1      9.017895         35   
1   2  60369.0  18.625000              2      7.766643         59   
2   3  55126.0  34.766667              1      3.632069          6   
3   4  51690.0  15.808333              1      5.372942         16   
4   5  28347.0  40.941667              3      3.822477         20   

   dist_unhealthy  age  pop_dense  retail_dense  ...  job_pz job_qp  job_qw  \
0       11.738935   49   0.885827      0.492126  ...       0      0       0   
1        6.805396   51   0.874016      0.342520  ...       0      0       0   
2       13.671772   44   0.944882      0.724409  ...       0      0       0   
3        4.333286   50   0.889764      0.444882  ...       0      0       0   
4        5.967121   38   0.744094      0.661417  ...       0      0       0   

   job_rn  job_sa  job_vv  job_zz  area_b  area_c  area_d  
0       0       0       1       0       0       1       0  
1       0       0       0       0       0       1       0  
2       0       0       0       0       0       1       0  
3       0       0       0       0       0       1       0  
4       0       0       0       0       0       0       1  

[5 rows x 47 columns]

#### Train and Test split:

`product` column is the target variable. `id` column is removed since it doesn't add value to the training process


In [17]:
# Split the data into features and targets before any preprocessing
x_columns = df.columns.drop(['product', 'id'])
y = pd.get_dummies(df['product']).values

X_train, X_test, y_train, y_test = train_test_split(df[x_columns], y, test_size=0.2, random_state=42)

### Step 4: Standardize and prepare data for modeling


In [18]:
# Apply imputation to numeric columns in train and use the same transformer for test
imputer = SimpleImputer(strategy='median')
X_train = imputer.fit_transform(X_train)
X_test = imputer.transform(X_test)

# Standardize the numeric columns
scaler = StandardScaler()
X_train_sc = scaler.fit_transform(X_train)
X_test_sc = scaler.transform(X_test)

### Step 5: Perform K-fold cross-validation and model training
- Train a model using K-fold cross-validation with 5 folds.
- The model consists of a sequential neural network with two hidden layers, using ReLU activation for the first hidden layer and L1 regularization for the second hidden layer.
- Dropout is applied to the first hidden layer to prevent overfitting.
- The model is trained using the Adam optimizer and categorical cross-entropy loss function.
- The accuracy of each fold is calculated and printed.

In [19]:
# Set up K-fold cross-validation
kf = KFold(5, shuffle=True, random_state=42)

for fold, (train_idx, test_idx) in enumerate(kf.split(X_train_sc), 1):

    print("X_train[train_idx] shape:", X_train_sc[train_idx].shape)
    print("y_train[train_idx] shape:", y_train[train_idx].shape)
    print("X_train[test_idx] shape:", X_train_sc[test_idx].shape)
    print("y_train[test_idx] shape:", y_train[test_idx].shape)
    print()


    print(f"Fold #{fold}")
    model = Sequential([
        Dense(50, input_dim=X_train_sc.shape[1], activation='relu'),
        Dropout(0.5),  # Dropout to prevent overfitting
        Dense(25, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
        Dense(y_train.shape[1], activation='softmax')
    ])

    model.compile(loss='categorical_crossentropy', optimizer='adam')

    # Train the model on the fold
    # Convert NumPy arrays to TensorFlow Tensors with float32 data type directly using appropriate indexes
    model.fit(tf.convert_to_tensor(X_train_sc[train_idx], dtype=tf.float32),
              tf.convert_to_tensor(y_train[train_idx], dtype=tf.float32),
              validation_data=(tf.convert_to_tensor(X_train_sc[test_idx], dtype=tf.float32),
                               tf.convert_to_tensor(y_train[test_idx], dtype=tf.float32)),
              epochs=50, batch_size=32, verbose=0)

    # Evaluate the model on the fold's test data
    predictions = model.predict(tf.convert_to_tensor(X_train_sc[test_idx], dtype=tf.float32))
    score = metrics.accuracy_score(np.argmax(y_train[test_idx], axis=1), np.argmax(predictions, axis=1))
    print(f"Accuracy for fold {fold}: {score:.4f}")


X_train[train_idx] shape: (1280, 45)
y_train[train_idx] shape: (1280, 7)
X_train[test_idx] shape: (320, 45)
y_train[test_idx] shape: (320, 7)

Fold #1


/home/jbenton/projects/sandbox/teach/teach/lib/python3.12/site-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
I0000 00:00:1757356158.958348  117960 cuda_executor.cc:1001] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node
Your kernel may have been built without NUMA support.
2025-09-08 13:29:19.248682: W tensorflow/core/common_runtime/gpu/gpu_device.cc:2343] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step 
Accuracy for fold 1: 0.7312
X_train[train_idx] shape: (1280, 45)
y_train[train_idx] shape: (1280, 7)
X_train[test_idx] shape: (320, 45)
y_train[test_idx] shape: (320, 7)

Fold #2


/home/jbenton/projects/sandbox/teach/teach/lib/python3.12/site-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step 
Accuracy for fold 2: 0.7312
X_train[train_idx] shape: (1280, 45)
y_train[train_idx] shape: (1280, 7)
X_train[test_idx] shape: (320, 45)
y_train[test_idx] shape: (320, 7)

Fold #3


/home/jbenton/projects/sandbox/teach/teach/lib/python3.12/site-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step 
Accuracy for fold 3: 0.6719
X_train[train_idx] shape: (1280, 45)
y_train[train_idx] shape: (1280, 7)
X_train[test_idx] shape: (320, 45)
y_train[test_idx] shape: (320, 7)

Fold #4


/home/jbenton/projects/sandbox/teach/teach/lib/python3.12/site-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step 
Accuracy for fold 4: 0.7281
X_train[train_idx] shape: (1280, 45)
y_train[train_idx] shape: (1280, 7)
X_train[test_idx] shape: (320, 45)
y_train[test_idx] shape: (320, 7)

Fold #5


/home/jbenton/projects/sandbox/teach/teach/lib/python3.12/site-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step 
Accuracy for fold 5: 0.6719


**Observation**
- The output shows the accuracy scores for each fold of the cross-validation process:

  - Fold scores: The output displays the fold number (for example, Fold #1) and the corresponding accuracy score (for example, 0.71) for each fold. The accuracy score represents the proportion of correctly predicted labels to the total number of labels in the test set. Higher accuracy scores indicate better performance of the model on the test data.

### Step 6: Final accuracy calculation on the test set

In [20]:
# Final evaluation on the standardized test set
final_predictions = model.predict(tf.convert_to_tensor(X_test_sc, dtype=tf.float32)) # Convert X_test_sc to a Tensor
final_score = metrics.accuracy_score(np.argmax(y_test, axis=1), np.argmax(final_predictions, axis=1))
print(f"Final accuracy on test set: {final_score:.4f}")


13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Final accuracy on test set: 0.6825



**Observation**
- The final accuracy score achieved by the model is 0.68.